bundler/lib/bundler/self_manager.rb

```diff
@@ -9,46 +9,48 @@ class SelfManager
     def restart_with_locked_bundler_if_needed
       return unless needs_switching? && installed?
 
-      restart_with_locked_bundler
+      restart_with(lockfile_version)
     end
 
     def install_locked_bundler_and_restart_with_it_if_needed
       return unless needs_switching?
 
-      install_and_restart_with_locked_bundler
+      Bundler.ui.info \
+        "Bundler #{current_version} is running, but your lockfile was generated with #{lockfile_version}. " \
+        "Installing Bundler #{lockfile_version} and restarting using that version."
+
+      install_and_restart_with(lockfile_version)
+    end
+
+    def update_bundler_and_restart_with_it_if_needed(target)
+      version = resolve_update_version_from(target)
+      return unless version
+
+      Bundler.ui.info "Updating bundler to #{version}."
+
+      install_and_restart_with(version)
     end
 
     private
 
-    def install_and_restart_with_locked_bundler
-      bundler_dep = Gem::Dependency.new("bundler", lockfile_version)
-      spec = fetch_spec_for(bundler_dep)
-      return if spec.nil?
+    def install_and_restart_with(version)
+      requirement = Gem::Requirement.new(version)
+      spec = find_latest_matching_spec(requirement)
 
-      Bundler.ui.info \
-        "Bundler #{current_version} is running, but your lockfile was generated with #{lockfile_version}. " \
-        "Installing Bundler #{lockfile_version} and restarting using that version."
+      if spec.nil?
+        Bundler.ui.warn "Your lockfile is locked to a version of bundler (#{lockfile_version}) that doesn't exist at https://rubygems.org/. Going on using #{current_version}"
+        return
+      end
 
       spec.source.install(spec)
     rescue StandardError => e
       Bundler.ui.trace e
       Bundler.ui.warn "There was an error installing the locked bundler version (#{lockfile_version}), rerun with the `--verbose` flag for more details. Going on using bundler #{current_version}."
```

Comment:
I wonder if we should now remove this rescue clause, given that the two errors we imagined could happen (invalid version in lockfile, and `https://rubygems.org` not in `Gem.sources`) are now either handled gracefully or no longer an issue.

I guess I'll leave this there for a few more versions and if nobody reports any issues remove it.